service/history/workflow/workflow_test/mutable_state_impl_test.go

```diff
@@ -0,0 +1,331 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package workflow_test contains tests for the workflow package. There are also tests in the workflow package itself,
+// but test packages force you to only test exported methods.
+// See https://github.com/maratori/testpackage#motivation for more on the rationale used here.
+package workflow_test
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	"go.temporal.io/api/history/v1"
+	taskqueuepb "go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/server/api/enums/v1"
+	historyspb "go.temporal.io/server/api/history/v1"
+	"go.temporal.io/server/api/historyservice/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/cluster"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"go.temporal.io/server/service/history/configs"
+	"go.temporal.io/server/service/history/events"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/tests"
+	"go.temporal.io/server/service/history/workflow"
+)
+
+func TestMutableStateImpl_ForceFlushBufferedEvents(t *testing.T) {
+	t.Parallel()
+
+	for _, tc := range []mutationTestCase{
+		{
+			name:              "signals<maxNumEvents",
+			signals:           1,
+			maxNumEvents:      2,
+			transactionPolicy: workflow.TransactionPolicyActive,
+			expectFailure:     false,
+		},
+		{
+			name:              "signals=maxNumEvents",
+			signals:           2,
+			maxNumEvents:      2,
+			transactionPolicy: workflow.TransactionPolicyActive,
+			expectFailure:     true,
+		},
+		{
+			name:              "signals>maxNumEvents",
+			signals:           3,
+			maxNumEvents:      2,
+			transactionPolicy: workflow.TransactionPolicyActive,
+			expectFailure:     true,
+		},
+	} {
+		t.Run(tc.name, tc.Run)
+	}
+}
+
+type mutationTestCase struct {
+	name              string
+	transactionPolicy workflow.TransactionPolicy
+	signals           int
+	maxNumEvents      int
+	expectFailure     bool
```

Comment:
nit: maybe `expectFlush` ? Fail inflight workflow task is just how the first step of flush.